ruby/command-t/ext/command-t/scanner.c

```diff
@@ -0,0 +1,181 @@
+// Copyright 2010-present Greg Hurrell. All rights reserved.
+// Licensed under the terms of the BSD 2-clause license.
+
+#include <ruby.h>
+#include <assert.h>
+#include <errno.h>
+#include <stdlib.h>  /* for qsort() */
+#include <string.h>  /* for strncmp() */
+#include <sys/mman.h>
+#include "scanner.h"
+#include "match.h"
+#include "matcher.h"
+#include "heap.h"
+#include "ext.h"
+#include "ruby_compat.h"
+
+// order matters; we want this to be evaluated only after ruby.h
+#ifdef HAVE_PTHREAD_H
+#include <pthread.h> /* for pthread_create, pthread_join etc */
+#endif
+
+typedef struct {
+    VALUE source;
+    size_t bufsize;
+    matches_t matches;
+} paths_t;
+
+void mark_paths(paths_t *paths) {
+    rb_gc_mark(paths->source);
+}
+
+void free_paths(paths_t *paths) {
+    munmap(paths, paths->bufsize);
+}
+
+VALUE CommandTPaths_from_array(VALUE klass, VALUE source) {
+    Check_Type(source, T_ARRAY);
+    rb_obj_freeze(source);
+
+    long len = RARRAY_LEN(source);
+    long bufsize = sizeof(paths_t) + len * sizeof(match_t);
+    paths_t *paths = mmap(NULL, bufsize,
```

Comment:
It has worked historically, but I personally haven't tested it on Windows for a long, long time (years). It's one of those things where I make a best effort to not break things, but ultimately we need a Windows-using maintainer to step forward if we want to really be on top of the Windows side, and in the absence of that then we just have to rely on user reports of stuff working (or not).